### ISSUE TITLE
Dropout in nn.MultiheadAttention Causing Attention Weight Sum > 1

### ISSUE THREAD

**Issue by 6etacat** [pytorch/pytorch#42929]
## 🐛 Bug

Attention weights sum to over 1 when dropout is used in MultiheadAttention. 

## To Reproduce

1. Start from the [official transformers tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html)
2. Use custom encoder layer derived from the official encoder layer to expose attention weights
3. Check attention weights while training
4. The first attention vector of the first batch should give you `[1.25, 0, 0, ...]`, and if you look at the sums, many of them are > 1. 
5. When using dropout=0, the sums will all be 1, which matches our expectations

Code to reproduce:
https://gist.github.com/6etacat/791a93d46382d8663736c9d27698f2be

## Environment

 - PyTorch Version (e.g., 1.0): 1.6.0
 - OS (e.g., Linux): Manjaro Linux
 - How you installed PyTorch (`conda`, `pip`, source): conca
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: RTX 2070
 - Any other relevant information: N/A

## Additional context

cc @albanD @mruberry @zhangguanheng66

**Comment by zhangguanheng66** [comment]
Can you put together a short code snippet so I can reproduce the results?

**Comment by 6etacat** [comment]
Running [this gist](https://gist.github.com/6etacat/791a93d46382d8663736c9d27698f2be) will reproduce the result. It is based on the transformer tutorial. I'll write a minimal version that will reproduce the issue later today.

**Comment by zhangguanheng66** [comment]
BTW, why do you expect the sum of attention weights to be 1? Any references I could study?

**Comment by 6etacat** [comment]
Yes. You can see the [code](https://github.com/pytorch/pytorch/blob/59f86923503813bf8e4ee290bf04253d1e37a881/torch/nn/functional.py#L4151) used to process. There is a softmax on line 4140. An average over a bunch of softmaxed vectors will still give you a sum to 1.

**Comment by 6etacat** [comment]
I'm also confused as why we want to output the average attention weights, but rather than attention weights of all the heads. There is value to visualizing what each head learns, and taking average is extremely simple.

**Comment by zhangguanheng66** [comment]
> I'm also confused as why we want to output the average attention weights, but rather than attention weights of all the heads. There is value to visualizing what each head learns, and taking average is extremely simple.

It was a decision made at the early stage of development, which is not the optimal. We realize the problem and in `MultiheadAttentionContainer`, we output non-average attention weights. see [here](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)

**Comment by zhangguanheng66** [comment]
> I'm also confused as why we want to output the average attention weights, but rather than attention weights of all the heads. There is value to visualizing what each head learns, and taking average is extremely simple.

I think, the the `softmax` layer is followed by `dropout` layer so I do expect a sum to be not 1.0. You can do a simple experimental like this:
```
import torch
a = torch.tensor([1., 2., 3.])
b = torch.nn.functional.softmax(a, dim=-1)
torch.nn.functional.dropout(b, p=0.5, training=True)
```
If you repeat the last line many times, you will see different sums.

**Comment by 6etacat** [comment]
Yes. Different sums are expected due to dropout, but dropout shouldn't change the values that are already there. Since we are dropping out a softmaxed tensor over `dim=-1`, we would expect that the value after dropout to still be between 0 and 1 and sum to an amount smaller equal to 1 due to softmax. However, this isn't the case:
![Screenshot from 2020-08-12 21-14-15](https://user-images.githubusercontent.com/15188392/90083718-4ca73400-dce1-11ea-8cd4-e3efe608b9c6.png)

You can clearly see that the first vector and the third vector of the attention weights, the first vector has a value of 1.25, which is clearly larger than 1, and the though the third vector has values smaller than 1, the sum is larger than 1.

**Comment by 6etacat** [comment]
> > I'm also confused as why we want to output the average attention weights, but rather than attention weights of all the heads. There is value to visualizing what each head learns, and taking average is extremely simple.
> 
> It was a decision made at the early stage of development, which is not the optimal. We realize the problem and in `MultiheadAttentionContainer`, we output non-average attention weights. see [here](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)

Ah. I did not know that there's a separate version of multiheaded attention in torchtext. I think we should mention that in the nn.MultiheadAttention documentation so people know where to look if they need to extract the weights of each head. 

Why did we create a new one? It also seems slightly less fully-featured than the one in torch.nn. Isn't it possible to make these changes to the main repo? Or is it because it's a breaking change we will have to wait until PyTorch 2.0?

**Comment by zhangguanheng66** [comment]
> > > I'm also confused as why we want to output the average attention weights, but rather than attention weights of all the heads. There is value to visualizing what each head learns, and taking average is extremely simple.
> > 
> > 
> > It was a decision made at the early stage of development, which is not the optimal. We realize the problem and in `MultiheadAttentionContainer`, we output non-average attention weights. see [here](https://github.com/pytorch/text/blob/master/torchtext/nn/modules/multiheadattention.py)
> 
> Ah. I did not know that there's a separate version of multiheaded attention in torchtext. I think we should mention that in the nn.MultiheadAttention documentation so people know where to look if they need to extract the weights of each head.
> 
> Why did we create a new one? It also seems slightly less fully-featured than the one in torch.nn. Isn't it possible to make these changes to the main repo? Or is it because it's a breaking change we will have to wait until PyTorch 2.0?

The container adds extra flexibility to MHA. People can try different attention mechanism there. We also add a few more broadcast support there (see relevant PRs for details).

**Comment by 6etacat** [comment]
I see. That makes sense. But I still think we should add a link from the nn.MultiheadAttention doc so it's easier to find. 

Any insights into why there would be 1.25 in the weights after dropout? It doesn't make sense to me.

**Comment by zhangguanheng66** [comment]
> I see. That makes sense. But I still think we should add a link from the nn.MultiheadAttention doc so it's easier to find.
> 
> Any insights into why there would be 1.25 in the weights after dropout? It doesn't make sense to me.

It seems that the behavior is related to `dropout`, not `MHA`. If you try the follow code snippet, you will see that the `dropout` func doesn't work that way.
```
import torch
a = torch.tensor([0.2, 0.2, 0.2, 0.2, 0.2])
torch.nn.functional.dropout(a, p=0.3, training=True)
```

**Comment by 6etacat** [comment]
I see. I didn't realize that dropout does normalization as well. Thanks for the patience explaining everything!

**Comment by parmarsuraj99** [comment]
Thanks for the insight, saved my day!
